repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java

```diff
@@ -992,35 +988,75 @@ public AtlasSearchResult directIndexSearch(SearchParams searchParams) throws Atl
+            ret.setAggregations(searchResult.indexQueryResult.getAggregationMap());
+            ret.setApproximateCount(searchResult.indexQuery.vertexTotals());
+        } catch (Exception e) {
+            LOG.error("Error while performing direct search for the params ({}), {}", searchParams, e.getMessage());
+            throw e;
+        }
+        return ret;
+    }
 
-            String indexName = getIndexName(params);
-
-            indexQuery = graph.elasticsearchQuery(indexName);
+    public List<AtlasVertex> directVerticesIndexSearch(SearchParams searchParams) throws AtlasBaseException {
+        IndexSearchParams params = (IndexSearchParams) searchParams;
+        setupRequestContext(params);
 
-            if (searchParams.getEnableFullRestriction()) {
-                addPreFiltersToSearchQuery(searchParams);
-            }
+        List<AtlasVertex> ret = new ArrayList<>();
 
-            AtlasPerfMetrics.MetricRecorder elasticSearchQueryMetric = RequestContext.get().startMetricRecord("elasticSearchQuery");
-            DirectIndexQueryResult indexQueryResult = indexQuery.vertices(searchParams);
-            if (indexQueryResult == null) {
+        try {
+            IndexSearchResult searchResult = performDirectIndexSearch(searchParams);
+            if (searchResult.indexQueryResult == null) {
                 return null;
             }
-            RequestContext.get().endMetricRecord(elasticSearchQueryMetric);
-            prepareSearchResult(ret, indexQueryResult, resultAttributes, true);
 
-            ret.setAggregations(indexQueryResult.getAggregationMap());
-            ret.setApproximateCount(indexQuery.vertexTotals());
+            Iterator<Result> iterator = searchResult.indexQueryResult.getIterator();
+            while (iterator.hasNext()) {
+                Result result = iterator.next();
+                AtlasVertex vertex = result.getVertex();
+                ret.add(vertex);
+            }
         } catch (Exception e) {
             LOG.error("Error while performing direct search for the params ({}), {}", searchParams, e.getMessage());
             throw e;
         }
         return ret;
     }
 
+    private void setupRequestContext(IndexSearchParams params) {
+        RequestContext.get().setRelationAttrsForSearch(params.getRelationAttributes());
+        RequestContext.get().setAllowDeletedRelationsIndexsearch(params.isAllowDeletedRelations());
+        RequestContext.get().setIncludeRelationshipAttributes(params.isIncludeRelationshipAttributes());
+    }
+
+    private record IndexSearchResult(AtlasIndexQuery indexQuery, DirectIndexQueryResult indexQueryResult) {
+    }
+
+    private IndexSearchResult performDirectIndexSearch(SearchParams searchParams) throws AtlasBaseException {
+        if(LOG.isDebugEnabled()){
+            LOG.debug("Performing ES search for the params ({})", searchParams);
+        }
+
+        IndexSearchParams params = (IndexSearchParams) searchParams;
```

Comment:
Let us not touch original `public AtlasSearchResult directIndexSearch(SearchParams searchParams)`
I do not want to see any regression or overhead for original API at all, it is fine if we have to duplicate a few lines for new method